pkg/github/github.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package github provides the functionality to send requests to the GitHub API.
+package github
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/google/go-github/v53/github"
+	"github.com/sethvargo/go-retry"
+	"golang.org/x/oauth2"
+)
+
+var ignoredStatusCodes = []int{400, 401, 403, 404, 422}
+
+type Config struct {
+	maxRetries        uint64
+	initialRetryDelay time.Duration
+	maxRetryDelay     time.Duration
+}
+
+// GitHub provides the minimum interface for sending requests to the GitHub API.
+type GitHub interface {
+	CreatePullRequestComment(ctx context.Context, owner, repo string, number int, body string) (int64, error)
+	UpdatePullRequestComment(ctx context.Context, owner, repo string, id int64, body string) error
+	DeletePullRequestComment(ctx context.Context, owner, repo string, id int64) error
+	ListPullRequestComments(ctx context.Context, owner, repo string, number int, opts *github.IssueListCommentsOptions) ([]*github.IssueComment, *github.Response, error)
+}
+
+var _ GitHub = (*GitHubClient)(nil)
+
+// GitHubClient implements the GitHub interface.
+type GitHubClient struct {
+	cfg    *Config
+	client *github.Client
+}
+
+// NewClient creates a new GitHub client.
+func NewClient(ctx context.Context, token string, opts ...Option) *GitHubClient {
+	cfg := &Config{
+		maxRetries:        3,
+		initialRetryDelay: 1 * time.Second,
+		maxRetryDelay:     20 * time.Second,
+	}
+
+	for _, opt := range opts {
+		if opt != nil {
+			cfg = opt(cfg)
+		}
+	}
+
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: token},
+	)
+	tc := oauth2.NewClient(ctx, ts)
```

Comment:
Is there an SDK function for "source the creds from the env"? That would let us support private key auth and token auth easily.

Reply:
Do you mean so we could auth as a user or as an app? It seems there are two different methods for this: https://github.com/google/go-github#authentication

Reply:
Yea, I'm thinking we might auth as an app in the future.

Reply:
We have: https://github.com/abcxyz/pkg/blob/main/githubapp/githubapp.go

We can wrap this, but I think we'll need to accept an HTTP Client and build a wrapper based on inputs to generate the HTTP client we want to use...doesnt seem like it exists.

Ill add an issue for this.

Reply:
Created #130 and added comment inline.